arcaflow_plugin_opensearch/opensearch_schema.py

```diff
@@ -1,30 +1,86 @@
 from dataclasses import dataclass
+import enum
 import typing
-from arcaflow_plugin_sdk import schema, validation
+from arcaflow_plugin_sdk import schema, validation, plugin
+
+
+class OperationType(enum.Enum):
+    CREATE = "create"
+    DELETE = "delete"
+    INDEX = "index"
+    UPDATE = "update"
+
+
+@dataclass
+class OperationMeta:
+    _index: typing.Annotated[
+        typing.Optional[str],
+        schema.name("index"),
+        schema.description(
+            "Optional name of the index that will receive the data. "
+            "If none is provided, the default index will be used."
+        ),
+    ] = None
+    _id: typing.Annotated[
+        typing.Optional[str],
+        schema.name("ID"),
+        schema.description("Optional ID for the data document."),
+    ] = None
 
 
 @dataclass
-class StoreDocumentRequest:
+class Operation:
+    operation: typing.Annotated[
+        # typing.Dict[OperationType, BulkUploadOperationMeta],
+        # Temporarily changing the key to a string in order to work
+        # around a workflow validation failure for the enum as a key
+        typing.Dict[str, OperationMeta],
+        schema.name("operation"),
+        schema.description("The operation type and associated operation metadata."),
+    ]
+
+
+@dataclass
+class BulkUploadObject(Operation):
+    data: typing.Annotated[
+        typing.Dict[str, typing.Any],
+        schema.name("data"),
+        schema.description("The JSON data document to upload to your index."),
+    ]
+
 
+bulk_upload_object_schema = plugin.build_object_schema(BulkUploadObject)
+
+
+@dataclass
+class BulkUploadList:
+    bulk_upload_list: typing.Annotated[
+        typing.List[BulkUploadObject],
+        schema.name("bulk upload list"),
+        schema.description("The list of objects for the bulk upload operation."),
+    ]
+
+
+@dataclass
+class DocumentRequest(BulkUploadList):
     url: typing.Annotated[
         str,
         schema.name("url"),
         schema.description("The URL for the Opensearch-compatible instance."),
     ]
-
-    index: typing.Annotated[
+    default_index: typing.Annotated[
         str,
         validation.min(1),
         schema.name("index"),
-        schema.description("Name of the index that will receive the data."),
-    ]
-
-    data: typing.Annotated[
-        typing.Dict[str, typing.Any],
-        schema.name("data"),
-        schema.description("Data to upload to your index"),
+        schema.description("Name of the default index that will receive the data."),
     ]
-
+    metadata: typing.Annotated[
+        typing.Optional[typing.Dict[str, typing.Any]],
+        schema.name("metadata"),
+        schema.description(
+            "Optional global metadata object that will be added " "to every document."
```

Comment:
Looks like the linter removed the code wrap.

Reply:
@dbutenhof I guess I'm not following. The `schema.name` is part of the self-documenting feature of the plugin. Often the name simply matches the key, but not always.